rfcs/syntax-safe-navigation-operator.md

```diff
@@ -0,0 +1,121 @@
+# Safe navigation postfix operator (?)
+
+## Summary
+
+Introduce syntax to navigate through `nil` values, or short-circuit with `nil` if it was encountered.
+
+
+## Motivation
+
+nil values are very common in Lua, and take care to prevent runtime errors. 
+
+Currently, attempting to index `dog.name` while caring for `dog` being nil requires some form of the following:
+
+```lua
+local dogName = nil
+if dog ~= nil then
+    dogName = dog.name
+end
+```
+
+...or the unusual to read...
+
+```lua
+local dogName = dog and dog.name
+```
+
+...which will return `false` if `dog` is `false`, instead of throwing an error because of the index of `false.name`.
+
+This trick gets worse in the case of calling methods. For example, let's suppose we wanted to call `dog.fetch()`, while `dog` is still potentially nil.
+
+The one-line `and` trick will no longer work (or will get less readable as you try to shoe-horn this in), so we must:
+
+```lua
+if dog ~= nil then
+    dog.fetch()
+end
+```
+
+But this gets even worse when it comes to chained indexing. Let's suppose we wanted to run `dog.owner.handshake()`, while `dog` can be nil and `owner` can be nil.
+
+```lua
+if dog ~= nil and dog.owner ~= nil then
+    dog.owner.handshake()
+end
+```
+
+...which gets even worse in the context of calling this function in, say, another function:
+
+```lua
+-- Oops! dog and dog.owner can be nil
+logDogName(getLogger(), dog.name, dog.owner:getDisplayName())
+
+-- In order to preserve this order (assuming argument execution order mattered)...
+local logger = getLogger()
+local name = dog and dog.name
+local displayName
+
+if dog.owner ~= nil then
+    displayName = dog.owner:getDisplayName()
+end
+
+logDogName(logger, name, displayName)
+```
+
+## Design
+
+The safe navigation operator will make all of these smooth, by supporting `x?.y` and similar indexing operators. `dog?.name` would resolve to `nil` if `dog` was nil, or the name otherwise. `owner?.handshake()` would only call `handshake` if `owner` is not nil.
+
+The long example would turn into:
+
+```lua
+logDogName(getLogger(), dog?.name, dog?.owner?:getDisplayName())
+```
+
+Failing the nil-safety check early would make the entire expression nil, for instance `dog?.body.legs` would resolve to `nil` if `dog` is nil, rather than resolve `dog?.body` into nil, then turning into `nil.legs`.
+
+The list of valid operators to follow the safe navigation operator would be:
+
+```lua
+dog?.name
+dog?.getName()

```

Comment:
Hmm. This would be equivalent to `(dog?.getName)()` under current/intuitive grammar. What was the intent here?

Reply:
The intent is for `getName` to be ran only if `dog` is not nil, which matches how a few languages, including TypeScript, interpret this code.

Reply:
The idea is the expression gets short circuited when the `?.` fails. That's why `a?.b.c.d.e` will work, even if `(a?.b).c.d.e` wouldn't (because `(a?.b)` is it's own expression, that gets replaced with nil, then indexed by c.d.e)

Reply:
What's the expression boundary though? E.g. `a?.b + 1` presumably doesn't short circuit?

Reply:
I'd have to read up on how an expression is expressed in the Luau parser, but no. 

I'm not really sure how to word this, or what ambiguities I'm missing.

Reply:
To find all that, we have prior art to help us along most of the way. It's also great because it helps to maintain some consistency in the behavior across languages.

Here's JavaScript's RFC on optional chaining: https://tc39.es/proposal-optional-chaining/

From some light reading, the reason why `t?.m()` automagically works in JavaScript is because they have `?.` start a different branch of the grammar that is identical to normal indexing syntax, but parses in a different way if `?.` is involved. Some expert in JS syntax will be able to confirm that.

For example, here's two call sites (I won't get into TemplateStringsArray because that's a red herring):

```
t.m`hi`;  // this is parsed as a function call that passes some TemplateStringsArray
t?.m`hi`; // this is illegal syntax
```

Reply:
Right. I think we should be explicit about the semantics of these operators in this RFC so that questions like "why can t?.x + 5 cause a runtime error while t?.x() can't" are obvious from how the RFC specifies the behavior.

Reply:
I'll write down in the RFC how binary operators like + operate, as well as how parentheses will operate. Is there anything else I should be thinking about?